canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -131,57 +133,138 @@ private[apiserver] final class ApiPackageManagementService private (
+          Future.fromTry(TypecheckUpgrades.typecheckUpgrades((pkgId1, pkg1), pkgId2, optPkg2)).recoverWith {
+            case err: UpgradeError =>
               Future.failed(
                 Validation.Upgradeability
-                  .Error(upgradingPackageId, upgradedPackageId, err)
+                  .Error(pkgId2, pkgId1, err)
                   .asGrpcError
               )
-
-            case Failure(err: Throwable) =>
+            case NonFatal(err) =>
               Future.failed(
                 InternalError
                   .Unhandled(
                     err,
                     Some(
-                      s"Typechecking upgrades for $upgradingPackageId failed with unknown error."
+                      s"Typechecking upgrades for $pkgId2 failed with unknown error."
                     ),
                   )
                   .asGrpcError
               )
           }
-        } yield ()
-      case None => {
-        logger.info(s"Package $upgradingPackageId does not upgrade anything.")
-        Future(())
       }
     }
   }
 
+  def typecheckUpgrades(typecheckPhase: TypecheckUpgrades.UploadPhaseCheck, optDar1: Option[(Ref.PackageId, Ast.Package)], optDar2: Option[(Ref.PackageId, Ast.Package)])(implicit
+                                                                                                                      loggingContext: LoggingContextWithTrace
+  ): Future[Unit] = {
+    (optDar1, optDar2) match {
+      case (None, _) | (_, None) =>
+        Future.unit
+
+      case (Some((pkgId1, pkg1)), Some((pkgId2, pkg2))) =>
+        strictTypecheckUpgrades(typecheckPhase, Some((pkgId1, pkg1)), pkgId2, Some(pkg2))
+    }
+  }
+
+  private def validateUpgrade(upgradingDar: Dar[(Ref.PackageId, Ast.Package)])(implicit
+      loggingContext: LoggingContextWithTrace
+  ): Future[Unit] = {
+    val (upgradingPackageId, upgradingPackage) = upgradingDar.main
+    val optUpgradingDar = Some((upgradingPackageId, upgradingPackage))
+    logger.info(s"Uploading DAR file for $upgradingPackageId in submission ID ${loggingContext.serializeFiltered("submissionId")}.")
+    existingVersionedPackageId(upgradingDar.main) match {
+      case Some(uploadedPackageId) =>
+        if (uploadedPackageId == upgradingPackageId) {
+          logger.info(s"Ignoring upload of package $upgradingPackageId as it has been previously uploaded")
+          Future.unit
+        } else {
+          logger.info(s"Package $upgradingPackageId can not be uploaded - $uploadedPackageId is already uploaded and at version ${upgradingPackage.metadata.version}")
+          Future.failed(Validation.UpgradeVersion.Error(uploadedPackageId, upgradingPackageId, upgradingPackage.metadata.version).asGrpcError)
+        }
+
+      case None =>
+        upgradingPackage.metadata.upgradedPackageId match {
+          case Some(upgradedPackageId) =>
+            for {
+              optUpgradedDar <- lookupDar(upgradedPackageId)
```

Comment:
Why are we continuing with a possible `None` here? Is there a situation where it's valid that the `metadata.upgradedPackageId` points to a non-existent DAR? If not, I'd simplify and fail early